### PR TITLE
Migrate torch fused patterns to common representation

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

Torch fused patterns migrated to new API

### Reason for changes

Unified processing of fused patterns allows to use common code in PTQ algorithms (MinMax)

### Tests
`test_pattern_manager` for Torch

### CODE REVIEW

File: nncf/torch/hardware/fused_patterns.py [hunk 1]
```diff
@@ -13,49 +30,124 @@
...
+    pattern = GraphPattern()
+    pattern.add_pattern_alternative(bn_act)
+    pattern.add_pattern_alternative(act_bn)
+    pattern.add_pattern_alternative(batch_norm)
+    pattern.add_pattern_alternative(activations)
+    return pattern
 
 
-PT_HW_FUSED_PATTERNS = _get_torch_hw_fused_patterns()
+def get_torch_hw_patterns():
```
Comment (KodiaqQ) [inline nncf/torch/hardware/fused_patterns.py:152]: Why do we need this method? Why not use `PatternsManager` as is?

File: nncf/torch/hardware/fused_patterns.py [hunk 2]
```diff
@@ -13,49 +30,124 @@
...
-    batch_norm_activations_permutation = batch_norm + activations | activations + batch_norm | batch_norm | activations
 
-    retval.register(linear_ops + arithmetic_ops, 'LINEAR + ARITHMETIC', match=True)
-    retval.register(linear_ops + batch_norm_activations_permutation, 'LINEAR + BN_ACT_PERM',
-                    match=True)
-    retval.register(batch_norm + activations, 'BN + ACTIVATIONS', match=True)
-    retval.register(activations + batch_norm, 'ACTIVATIONS + BN', match=True)
-    retval.register(arithmetic_ops + batch_norm_activations_permutation,
-                    'ARITHMETIC + BN_ACT_PERM', match=True)
+def batch_norm_activations_permutations():
```
Comment (KodiaqQ) [inline nncf/torch/hardware/fused_patterns.py:132]: My suggestion is not to add so-called `permutations` of the patterns. This method contains the list of existing combinations such as `batch norm + activations`, `activations + batch norm` and etc., that should be described as BATCH_NORM_ACTIVATIONS and corresponding patterns.

File: nncf/torch/hardware/fused_patterns.py (hunk 1 above)
Comment (daniil-lyakhov) [reply nncf/torch/hardware/fused_patterns.py:152]: I did it to simplify code inside torch algorithms: instead of importing BackendType, TargetDevice  and using them every time I made a default version. I believe this approach is easier to understand as Torch has no target specific patterns and logic yet

**Comment by alexsu52** [comment]
@KodiaqQ, please, review.

File: nncf/torch/hardware/fused_patterns.py (hunk 2 above)
Comment (daniil-lyakhov) [reply nncf/torch/hardware/fused_patterns.py]: Done, please check

File: nncf/experimental/torch/search_building_blocks/search_graph.py [hunk 3]
```diff
@@ -261,7 +263,8 @@ def get_merged_original_graph_with_pattern(orig_graph: nx.DiGraph, hw_fused_ops:
     if not hw_fused_ops:
         return merged_graph
     # pylint: disable=protected-access
-    pattern_fusing_graph = PT_HW_FUSED_PATTERNS.get_full_pattern_graph()
+    pattern_fusing_graph =\
+        PatternsManager().get_full_pattern_graph(BackendType.TORCH, TargetDevice.ANY)
```
Comment (KodiaqQ) [inline nncf/experimental/torch/search_building_blocks/search_graph.py:267]: Can these patterns be applied only for ANY (CPU) device?

File: nncf/torch/hardware/fused_patterns.py [hunk 4]
```diff
@@ -13,49 +27,148 @@
...
-    h_swish = create_h_swish_act()
-    activations = atomic_activations | swish | h_swish | h_sigmoid
-    retval.register(activations, 'ACTIVATIONS', match=False)
 
-    arithmetic_ops = GraphPattern()
-    arithmetic_ops.add_node(**ARITHMETIC_OPERATIONS)
-    retval.register(arithmetic_ops, ARITHMETIC_OPERATIONS['label'], match=False)
+@PT_HW_FUSED_PATTERNS.register(PatternNames.L2_NORM)
+def create_l2_norm_operations():
+    return create_l2_norm()
```
Comment (KodiaqQ) [inline nncf/torch/hardware/fused_patterns.py:37]: Can we reduce the number of files and replace methods from the `nncf.torch.graph.patterns` directly here? Since they are not used anymore, but only here.

File: nncf/torch/graph/pattern_operations.py [hunk 5]
```diff
@@ -11,68 +11,69 @@
  limitations under the License.
 """
 from nncf.common.graph.patterns import merge_two_types_of_operations
+from nncf.common.graph.patterns import GraphPattern
 
-LINEAR_OPERATIONS = {'type': ['linear',
```
Comment (kshpv) [inline nncf/torch/graph/pattern_operations.py:15]: I think we should update patterns on Metatypes in this PR how it did in ONNX and OV

**Comment by kshpv** [comment]
I would suggest to add @vshampor to review torch patterns

File: nncf/common/graph/patterns/patterns.py [hunk 6]
```diff
@@ -281,6 +281,9 @@ class PatternNames(Enum):
     and would not be quantized in compression pipeline.
     """
 
+    # ATOMIC OPERATIONS
+    L2_NORM = PatternDesc('l2_norm')
```
Comment (kshpv) [inline nncf/common/graph/patterns/patterns.py:285]: Can we somehow divide Torch specific patterns from common patterns?

**Review by kshpv — CHANGES_REQUESTED** [review]

File: nncf/torch/quantization/algo.py [hunk 7]
```diff
@@ -347,7 +349,8 @@ def __init__(self, quant_config: Dict, target_model: NNCFNetwork,
                          precision_init_params, range_init_params,
                          hw_config)
 
-        self._pattern_fusing_graph = PT_HW_FUSED_PATTERNS.get_full_pattern_graph()
+        self._pattern_fusing_graph =\
+            PatternsManager().get_full_pattern_graph(BackendType.TORCH, TargetDevice.ANY)
```
Comment (KodiaqQ) [inline nncf/torch/quantization/algo.py:353]: Please, use the predefined `TargetDevice`, since the quantization is possible not only for `ANY` device.

File: nncf/common/graph/patterns/patterns.py (hunk 6 above)
Comment (kshpv) [inline nncf/common/graph/patterns/patterns.py:285]: We already have NORMALIZE_L2_MULTIPLY. What is the diff between them?

File: nncf/torch/graph/pattern_operations.py (hunk 5 above)
Comment (daniil-lyakhov) [reply nncf/torch/graph/pattern_operations.py:15]: The problem is that in contrast with ONNX backend, Torch backend uses operation names instead of metatypes for pattern matching. I suggest to cover this in a separate PR

File: nncf/experimental/torch/search_building_blocks/search_graph.py (hunk 3 above)
Comment (daniil-lyakhov) [reply nncf/experimental/torch/search_building_blocks/search_graph.py:267]: Torch has no device specific logic in patterns

File: nncf/common/graph/patterns/patterns.py (hunk 6 above)
Comment (daniil-lyakhov) [reply nncf/common/graph/patterns/patterns.py:285]: I'll check it

File: tests/torch/test_nncf_network.py [hunk 8]
```diff
@@ -622,7 +624,8 @@ def forward(self, input_):
     def test_get_ip_graph_with_merged_operations(self, mock_graph_factory, dot_file_name):
         mock_graph = mock_graph_factory()
         ip_graph = get_ip_graph_for_test(mock_graph)
-        pattern = PT_HW_FUSED_PATTERNS.get_full_pattern_graph()
+        pattern =\
+            PatternsManager().get_full_pattern_graph(BackendType.TORCH, TargetDevice.ANY)
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py:628]: I copied it from MinMax, I'll check

File: nncf/torch/quantization/metrics.py [hunk 9]
```diff
@@ -329,7 +331,8 @@ def _marking_edges(self, graph, node_key, queue, mark=True):
                 self.stats.quantized_edges_in_cfg += 1
 
     def get_merged_original_graph_with_patterns(self, original_graph: PTNNCFGraph):
-        pattern = PT_HW_FUSED_PATTERNS.get_full_pattern_graph()
+        pattern =\
+            PatternsManager().get_full_pattern_graph(BackendType.TORCH, TargetDevice.ANY)
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/quantization/metrics.py:335]: I afraid that device param wasn't indented to be here, I probably need to refactor whole quantization algo to pass this param which at the end will change nothing. My suggestion is: let's refactor this when device specific patterns will be introduced in Torch backend

File: nncf/torch/quantization/algo.py (hunk 7 above)
Comment (daniil-lyakhov) [reply nncf/torch/quantization/algo.py:353]: Please see https://github.com/openvinotoolkit/nncf/pull/1566/files/01b9ab9f9143ec0361290f5bb322521e4bdf0f83#r1105622381

File: nncf/experimental/torch/search_building_blocks/search_graph.py (hunk 3 above)
Comment (KodiaqQ) [reply nncf/experimental/torch/search_building_blocks/search_graph.py:267]: > Torch has no device specific logic in patterns

Do you mean quantization in general or some specific algorithms that use this method?

**Comment by daniil-lyakhov** [comment]
> I would suggest to add @vshampor to review torch patterns

This code doesn't change any logic inside torch algorithms, it's just using different representation. I suggest to keep this PR as small as possible, and when some fussed pattern logic will have to be hanged we will refactor this code

File: nncf/torch/quantization/metrics.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply nncf/torch/quantization/metrics.py:335]: But the `MinMaxQuantization` algorithm has a `device` as an option. Or are these files not used in the `MinMaxQuantization` pipeline?

File: nncf/torch/graph/pattern_operations.py (hunk 5 above)
Comment (kshpv) [reply nncf/torch/graph/pattern_operations.py:15]: I know it, but the think that you need to do is only to change the name of the layer to Metatype.

**Comment by kshpv** [comment]
> > I would suggest to add @vshampor to review torch patterns
> 
> This code doesn't change any logic inside torch algorithms, it's just using different representation. I suggest to keep this PR as small as possible, and when some fussed pattern logic will have to be hanged we will refactor this code

As @vshampor added most of the patterns, I suggest adding him to double check that all patterns are registered

File: nncf/experimental/torch/search_building_blocks/search_graph.py (hunk 3 above)
Comment (daniil-lyakhov) [reply nncf/experimental/torch/search_building_blocks/search_graph.py:267]: There are no algorithms in NNCF Torch that's using information about device when collecting fussing pattern. Fussing patterns are always the same

Comment (KodiaqQ) [reply nncf/experimental/torch/search_building_blocks/search_graph.py:267]: @alexsu52, @AlexKoff88, is the statement above correct? Should PyTorch PTQ quantization use types of devices that differ from ANY?

File: nncf/torch/quantization/metrics.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/quantization/metrics.py:335]: This PR is not related to MinMaxQuantization pipeline, this PR converts existing patterns to a new representation.
Besides no matter device Torch pattern manager will return the same set of patterns, so it's safe to use in `MinMaxQuantization`

File: nncf/common/graph/patterns/patterns.py (hunk 6 above)
Comment (daniil-lyakhov) [reply nncf/common/graph/patterns/patterns.py:285]: It's differs from Openvino `NORMALIZE_L2_MULTIPLY`

File: nncf/torch/graph/pattern_operations.py (hunk 5 above)
Comment (daniil-lyakhov) [reply nncf/torch/graph/pattern_operations.py:15]: What do you mean by the new API has patterns only on Metatypes? Could you please show me this new API?

File: tests/torch/test_nncf_network.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py]: Done

File: nncf/torch/hardware/fused_patterns.py (hunk 4 above)
Comment (daniil-lyakhov) [reply nncf/torch/hardware/fused_patterns.py]: Done, please check

File: nncf/torch/graph/pattern_operations.py (hunk 5 above)
Comment (daniil-lyakhov) [reply nncf/torch/graph/pattern_operations.py:15]: I want to use torch patterns inside `MinMaxQuantization`, and for that I don't need to change types here. I would do this in a separate PR with pattern matching algorithm refactoring

Comment (kshpv) [reply nncf/torch/graph/pattern_operations.py:15]: Please, look at this PR - https://github.com/openvinotoolkit/nncf/commit/54540ba1e1f50ad961c8cb3e2b285f361b332ac7#diff-62c24dee1521c104745fab8ceee4a346d952825ea04e70152033a1ceb78265afL16
This is how the migration of patterns was done for ONNX.

All matching logic already is implemented.  About what refactoring did you write?

Comment (kshpv) [reply nncf/torch/graph/pattern_operations.py:15]: > This method, for example, could be simplified when all backends will be migrated on metatypes https://github.com/openvinotoolkit/nncf/blob/develop/nncf/common/graph/graph_matching.py#L76-L95

This is true, but it should be done only when ALL backends are migrated. If we add only Torch I don't see any refactoring in matching logic

**Review by kshpv — APPROVED** [review]

File: nncf/experimental/torch/search_building_blocks/search_graph.py [hunk 10]
```diff
@@ -261,7 +263,8 @@ def get_merged_original_graph_with_pattern(orig_graph: nx.DiGraph, hw_fused_ops:
     if not hw_fused_ops:
         return merged_graph
     # pylint: disable=protected-access
-    pattern_fusing_graph = PT_HW_FUSED_PATTERNS.get_full_pattern_graph()
+    pattern_fusing_graph =\
+        PatternsManager.get_full_pattern_graph(BackendType.TORCH, TargetDevice.ANY)
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/search_building_blocks/search_graph.py:267]: I suggest to set proper target device here in a separate PR

Comment (daniil-lyakhov) [reply nncf/experimental/torch/search_building_blocks/search_graph.py:267]: CC: @ljaljushkin

**Review by KodiaqQ — APPROVED** [review]

**Review by vshampor — APPROVED** [review]
